Fix handling of loop alignment targets

BruceForstall opened the pull request:
Now that `BBJ_COND` block false target is not necessarily fall-through, the loop alignment code needs to consider the false target as what induces a loop via a false target back edge.

Also, stop marking blocks after possible loop back edges for aligned loops as needing labels; they do not need to be so marked.

BruceForstall committed:
Fix handling of loop alignment targets

Now that `BBJ_COND` block false target is not necessarily
fall-through, the loop alignment code needs to consider the
false target as what induces a loop via a false target
back edge.

Also, stop marking blocks after possible loop back edges for
aligned loops as needing labels; they do not need to be so marked.

BruceForstall commented:
fyi, I found this with `JitOptRepeat`

BruceForstall commented:
@kunalspathak @amanasifkhalid PTAL
cc @dotnet/jit-contrib

amanasifkhalid reviewed (approved):
CodeGen changes LGTM; I'll leave the emitter-level changes to Kunal. I look forward to seeing what else the flowgraph churn has messed with...

kunalspathak commented:
> Also, stop marking blocks after possible loop back edges for aligned loops as needing labels; they do not need to be so marked.

Can you remind me why this tracking is no longer needed?

BruceForstall commented:
> Can you remind me why this tracking is no longer needed?

I was looking for cases that use `ehCodeOffset`, and there aren't any. What I didn't realize is that without the label, we can have a loop bottom (an IG where we set `igLoopBackEdge`), where we continue to add instructions after the backwards branch. That means the IG size increases and `getLoopSize` can return a larger value, thus some loops will get rejected for alignment because they are too large.

I'll restore this label code with a comment to this effect. I'll also only set the label if `igLoopBackEdge` was set.

BruceForstall committed:
Restore setting label on BasicBlock after setting `igLoopBackEdge`

This minimizes the size of an aligned loop by `getLoopSize`.

Also, add `superpmi.py asmdiffs --alignloops` argument to allow
specifying that you want to run asmdiffs with loop alignment enabled.
By default, asmdiffs is done with loop alignment disabled, to avoid
too many extraneous diffs.

BruceForstall commented:
This change does lead to new aligned loops in some cases.

BruceForstall commented:
Here are sumary diffs on win-x64 with `DOTNET_JitAlignLoops=1`:

Diffs are based on <span style="color:#1460aa">2,492,126</span> contexts (<span style="color:#1460aa">977,958</span> MinOpts, <span style="color:#1460aa">1,514,168</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">6,660 (0.27%)</span>


<details>
<summary>Overall (<span style="color:red">+1,369</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|41,770,334|<span style="color:red">+307</span>|
|benchmarks.run.windows.x64.checked.mch|8,671,738|<span style="color:red">+29</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|35,201,604|<span style="color:red">+119</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|12,689,725|<span style="color:red">+3</span>|
|coreclr_tests.run.windows.x64.checked.mch|392,760,156|<span style="color:red">+123</span>|
|libraries.pmi.windows.x64.checked.mch|61,143,010|<span style="color:red">+84</span>|
|libraries_tests.run.windows.x64.Release.mch|276,208,060|<span style="color:red">+475</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|133,750,789|<span style="color:red">+138</span>|
|realworld.run.windows.x64.checked.mch|14,193,741|<span style="color:red">+47</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,074,699|<span style="color:red">+44</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:red">+1,369</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|
|---|--:|--:|
|aspnet.run.windows.x64.checked.mch|27,111,609|<span style="color:red">+307</span>|
|benchmarks.run.windows.x64.checked.mch|8,671,377|<span style="color:red">+29</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|20,963,356|<span style="color:red">+119</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,515,026|<span style="color:red">+3</span>|
|coreclr_tests.run.windows.x64.checked.mch|119,245,778|<span style="color:red">+123</span>|
|libraries.pmi.windows.x64.checked.mch|61,029,491|<span style="color:red">+84</span>|
|libraries_tests.run.windows.x64.Release.mch|100,715,823|<span style="color:red">+475</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|122,939,939|<span style="color:red">+138</span>|
|realworld.run.windows.x64.checked.mch|13,804,036|<span style="color:red">+47</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|5,073,790|<span style="color:red">+44</span>|


</div></details>

BruceForstall commented:
@kunalspathak This is ready, with updates for feedback.

Note to all: I added a `superpmi.py asmdiffs --alignloops` option that allows specifying to do asmdiffs with loop alignment enabled.

kunalspathak commented on src/coreclr/scripts/superpmi.py:
```diff
@@ -1971,17 +1972,19 @@ def replay_with_asm_diffs(self):
             "DOTNET_JitDisasm": "*",
             "DOTNET_JitUnwindDump": "*",
             "DOTNET_JitEHDump": "*",
-            "DOTNET_JitDiffableDasm": "1", # to be removed
```
remember to undo this.

BruceForstall replied:
Why? This is the obsolete switch which has been replaced with JitDisasmDiffable, which is still set.

kunalspathak replied:
I misread your comment about "# to be removed" thinking that you want to revert this change before merging.

kunalspathak commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -2588,7 +2586,6 @@ void CodeGen::genReportEH()
                 }
                 else
                 {
-                    assert(bbLabel->bbEmitCookie != nullptr);
```
why is this removed?

BruceForstall replied:
The call to `ehCodeOffset` has the same assert, so it didn't seem necessary.

kunalspathak reviewed (approved):
LGTM. Thanks!
